modules/thermal_hydraulics/test/tests/controls/delay_control/gold/test_out.csv

```diff
@@ -2,7 +2,7 @@ time,p0_inlet,p0_inlet_delayed
 0,100000,100000
 0.1,100100,100000
 0.2,100200,100000
-0.3,100200,100000
+0.3,100200,5.5511151231258e-11
```

Comment:
@joshuahansel

Reply:
I m suspicious of this. Pressure should not jump from 1e5 to 0

Reply:
That is an excellent point 😆

Reply:
Yes, definitely don't accept this change. I discussed this test above. @YaqiWang did you try the change to `THMControl` that I suggested? If that doesn't fix the test, let me know.

Reply:
(Later I may try the change to `THMControl` in a separate PR, but it shouldn't hold up the show here.)

Reply:
The test fails for a different reason than I thought (or maybe it fails for two reasons). The problem I see now is that the control depends on a PP (`p0_inlet`), but the control is now executing before it and adding a zero point because the PP is still zero. I'm not sure what the right solution is yet.

Reply:
It's definitely a requirement that our controls be able to take PPs as input - we need to control components based on sensors, for example. It's unfortunate that we don't have cross-system dependency resolution - it's the source of a lot of headaches like this. As for executing all controls after all PPs, I *think* the case of controlling a PP value is more rare, but I'm sure it's there.

Reply:
Sorry @joshuahansel I have not tried. I was busy on something else. I guess @lindsayad was interested in this possibly due to the refactoring of setup functions in FEProblemBase.

Reply:
One solution is to create an execute flag like `EXEC_PRE_CONTROL`, then we can tag whatever objects including pps to that flag. If we have to make it happen on initial only, the flag can be `EXEC_PRE_INITIAL_CONTROL`.

Reply:
The PID control should be affected too, not sure why it is not. 
It is based on pp values

Reply:
Yeah the regression test for it appears to be dumb and is only checking the final solution to `rel_err=1e-4`. There's a heavy test that uses it that is apparently passing though.

Reply:
hey I wrote that test haha
yeah it s probably my fault it s so lenient

EXEC_PRE_CONTROL is not the preferred option here

Reply:
> hey I wrote that test haha

Nope, it was me and David. You wrote the one in the framework :)

Since `EXEC_PRE_CONTROL` is not preferred, do we have any idea what we could do about this?

Reply:
I was talking about PIDTransientControl I guess. I d expect it to diff here. 
Anyhow. 

There's a few paths forward:
- shelf this and use it as more motivation for general dependency resolution. There is a lot to gain from working on this.
- work with @YaqiWang to make his needs work in Griffin without this. I think I remember that he is trying to avoid providing initial conditions to an object, letting the control handle the initialization. Depending on whether this is wrapped in an action already, it might be tolerable to provide ICs programmatically. We can invite you next week to the meeting

Reply:
I think automatic general dependency resolution is an idea. I do not know the MOOSE team plan. But it is a high risk task requiring huge effort without too much gain on new capabilities.

@joshuahansel I am wondering whether you have similar issue on every time steps? Are you guys running controls at time step end for the next time step? Another solution is to call `executeControls` twice on initial. The one before `execute(EXEC_INITIAL)` will only execute controls that are independent on results of other moose objects. We can add a private parameter in `Control` so that we can let developers determine which catalog a control belongs to.

Reply:
I have one more idea that I'm going to try: having `THMControl` execute only on `TIMESTEP_BEGIN` instead of on both `INITIAL` and `TIMESTEP_BEGIN`. I cannot think of any use THM has for executing on both `INITIAL` and `TIMESTEP_BEGIN`. This would stagger the execution of our controls and PPs, since for PPs, we usually do `INITIAL` and `TIMESTEP_END`. I'll let you know how it goes.

Reply:
It works - I pushed it up. Sorry for the long debate. Hopefully this is the end of it!

Reply:
The THM heavy test and the Sockeye test can both be re-golded.